Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php

```diff
@@ -0,0 +1,127 @@
+<?php
+namespace Neos\Flow\Persistence\Doctrine;
+
+/*
+ * This file is part of the Neos.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Doctrine\ORM\Event\OnFlushEventArgs;
+use Doctrine\ORM\UnitOfWork;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Persistence\Exception\ObjectValidationFailedException;
+use Neos\Flow\Persistence\PersistenceManagerInterface;
+use Neos\Flow\Reflection\ClassSchema;
+use Neos\Utility\ObjectAccess;
+use Neos\Flow\Reflection\ReflectionService;
+use Neos\Utility\TypeHandling;
+use Neos\Flow\Validation\ValidatorResolver;
+
+/**
+ * Flow's Doctrine PersistenceManager
+ *
+ * @Flow\Scope("singleton")
+ * @api
+ */
+class ObjectValidationListener
+{
+    /**
+     * @Flow\Inject
+     * @var ObjectManager
+     */
+    protected $entityManager;
+
+    /**
+     * @Flow\Inject
+     * @var ValidatorResolver
+     */
+    protected $validatorResolver;
+
+    /**
+     * @Flow\Inject
+     * @var ReflectionService
+     */
+    protected $reflectionService;
+
+    /**
+     * @Flow\Inject
+     * @var PersistenceManagerInterface
+     */
+    protected $persistenceManager;
+
+    /**
+     * An onFlush event listener used to validate entities upon persistence.
+     *
+     * @param OnFlushEventArgs $eventArgs
+     * @return void
+     * @throws ObjectValidationFailedException
+     */
+    public function onFlush(OnFlushEventArgs $eventArgs)
+    {
+        /** @var UnitOfWork $unitOfWork */
+        $unitOfWork = $this->entityManager->getUnitOfWork();
+        $entityInsertions = $unitOfWork->getScheduledEntityInsertions();
+
+        $validatedInstancesContainer = new \SplObjectStorage();
+        $knownValueObjects = [];
+        foreach ($entityInsertions as $entity) {
+            $className = TypeHandling::getTypeForValue($entity);
+            if ($this->reflectionService->getClassSchema($className)->getModelType() === ClassSchema::MODELTYPE_VALUEOBJECT) {
+                $identifier = $this->persistenceManager->getIdentifierByObject($entity);
+
+                if (isset($knownValueObjects[$className][$identifier]) || $unitOfWork->getEntityPersister($className)->exists($entity)) {
+                    unset($entityInsertions[spl_object_hash($entity)]);
+                    continue;
+                }
+
+                $knownValueObjects[$className][$identifier] = true;
+            }
+            $this->validateObject($entity, $validatedInstancesContainer);
+        }
+
+        ObjectAccess::setProperty($unitOfWork, 'entityInsertions', $entityInsertions, true);
```

Comment:
TIL. Didn't know we did this kind of nasty magic there... But probably no other way currently?

Reply:
TBH I don't know. We sure did it for a reson back then…